src/goLanguageServer.ts

```diff
@@ -47,38 +58,63 @@ let languageServerDisposable: vscode.Disposable;
 let latestConfig: LanguageServerConfig;
 let serverOutputChannel: vscode.OutputChannel;
 
-// startLanguageServer starts the language server (if enabled), returning
-// true on success.
-export async function registerLanguageFeatures(ctx: vscode.ExtensionContext): Promise<boolean> {
-	// Subscribe to notifications for changes to the configuration of the language server.
-	ctx.subscriptions.push(vscode.workspace.onDidChangeConfiguration((e) => watchLanguageServerConfiguration(e)));
+// defaultLanguageProviders is the list of providers currently registered.
+let defaultLanguageProviders: vscode.Disposable[] = [];
+
+// restartCommand is the command used by the user to restart the language
+// server.
+let restartCommand: vscode.Disposable;
+
+// restartLanguageServer wraps all of the logic needed to restart the
+// language server. It can be used to enable, disable, or otherwise change
+// the configuration of the server.
+export let restartLanguageServer: () => {};
+
+// registerLanguageFeatures starts the language server (if enabled).
+// It registers the default language providers if the server is disabled
+// or not possible to start.
+export async function registerLanguageFeatures(ctx: vscode.ExtensionContext) {
+	// Subscribe to notifications for changes to the configuration
+	// of the language server, even if it's not currently in use.
+	ctx.subscriptions.push(vscode.workspace.onDidChangeConfiguration(
+		(e) => watchLanguageServerConfiguration(e)
+	));
+
+	// Set the function that is used to restart the language server.
+	restartLanguageServer = async () => {
+		startLanguageServerWithFallback(ctx, false);
+	};
+	startLanguageServerWithFallback(ctx, true);
+}
 
-	const config = parseLanguageServerConfig();
-	if (!config.enabled) {
-		return false;
+// startLanguageServerWithFallback starts the language server, if enabled,
+// or falls back to the default language providers.
+async function startLanguageServerWithFallback(ctx: vscode.ExtensionContext, activation: boolean) {
+	const cfg = parseLanguageServerConfig();
+	if (cfg.enabled) {
+		// If the language server is gopls, we can check if the user needs to
+		// update their gopls version. We do this only once per VS Code
+		// activation to avoid inundating the user.
+		if (activation && cfg.serverName === 'gopls') {
+			const tool = getTool(cfg.serverName);
+			if (!tool) {
+				return false;
```

Comment:
It's to get a Tool object to pass into shouldUpdateLanguageServer - not ideal, but it's a result of the current state because I had wanted to change the tool installation logic but never finished working on it. I renamed the `parseLanguageServerConfig` object.